vault/eventbus/filter.go

```diff
@@ -4,43 +4,62 @@
 package eventbus
 
 import (
+	"context"
+	"fmt"
 	"slices"
 	"sort"
+	"strings"
 	"sync"
 
 	"github.com/hashicorp/vault/helper/namespace"
 	"github.com/hashicorp/vault/sdk/logical"
 	"github.com/ryanuber/go-glob"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-// Filters keeps track of all the event patterns that each node is interested in.
+const globalCluster = ""
+
+// Filters keeps track of all the event patterns that each cluster is interested in.
 type Filters struct {
 	lock    sync.RWMutex
-	self    nodeID
-	filters map[nodeID]*NodeFilter
+	self    clusterID
+	filters map[clusterID]*ClusterFilter
+
+	// notifyChanges is used to notify about changes to filters. The condition variables are tied to single lock above.
+	notifyChanges map[clusterID]*sync.Cond
 }
 
-// nodeID is used to syntactically indicate that the string is a node's name identifier.
-type nodeID string
+// clusterID is used to syntactically indicate that the string is a cluster's identifier.
+type clusterID string
 
 // pattern is used to represent one or more combinations of patterns
 type pattern struct {
+	namespacePatterns string // space-separated (spaces are not allowed in namespaces, and slices are not comparable)
 	eventTypePattern  string
-	namespacePatterns []string
 }
 
-// NodeFilter keeps track of all patterns that a particular node is interested in.
-type NodeFilter struct {
-	patterns []pattern
+func (p pattern) String() string {
+	return fmt.Sprintf("{ns=%s,ev=%s}", p.namespacePatterns, p.eventTypePattern)
+}
+
+func (p pattern) isEmpty() bool {
+	return p.namespacePatterns == "" && p.eventTypePattern == ""
+}
+
+// ClusterFilter keeps track of all patterns that a particular cluster is interested in.
+type ClusterFilter struct {
+	patterns sets.Set[pattern]
 }
 
-func (nf *NodeFilter) match(ns *namespace.Namespace, eventType logical.EventType) bool {
+// match checks if the given ns and eventType matches any pattern in the cluster's filter.
+// Must be called while holding a (read) lock for the filter.
+func (nf *ClusterFilter) match(ns *namespace.Namespace, eventType logical.EventType) bool {
 	if nf == nil {
 		return false
 	}
-	for _, p := range nf.patterns {
+	for p := range nf.patterns {
 		if glob.Glob(p.eventTypePattern, string(eventType)) {
-			for _, nsp := range p.namespacePatterns {
+			for _, nsp := range strings.Split(p.namespacePatterns, " ") {
```

Comment:
I'll clean up the logic a little here. We do have to be careful because the root namespace has path `""`, and we do want to be able to match on that as a namespace.